ykrt/src/thread_intercept.rs

```diff
@@ -1,11 +1,54 @@
 use std::os::raw::{c_int, c_void};
 
+use std::{ffi::CString, ptr::null_mut};
+
+use libc::{dlsym, free, malloc, pthread_attr_t, pthread_create, pthread_t};
+
+const SHADOW_STACK_SIZE: usize = 1000000;
+
+#[derive(Debug)]
+struct ThreadRoutine {
+    pub args: *mut c_void,
+    pub start_routine: extern "C" fn(*mut c_void) -> *mut c_void,
+}
+
+extern "C" fn wrap_thread_routine(arg: *mut c_void) -> *mut c_void {
+    let str = CString::new("shadowstack_0").unwrap();
+    unsafe {
+        //  Obtain address of a shadowstack_0 symbol
+        let shadowstack_addr: *mut libc::c_void = dlsym(null_mut(), str.as_ptr() as *const i8);
+        if shadowstack_addr.is_null() {
+            panic!("Unable to find shadowstack address!")
+        }
+        // Allocate stack
+        let stack_addr = malloc(SHADOW_STACK_SIZE);
+        if stack_addr.is_null() {
+            panic!("Unable allocate stack!")
+        }
+        // Set shadowstack addr with new allocated stack
+        *(shadowstack_addr as *mut *mut c_void) = stack_addr;
+        let thread_routine: &ThreadRoutine = (arg as *mut ThreadRoutine)
+            .as_ref()
+            .expect("Thread routine function and args data.");
+        // Call original thread routine
+        let result = (thread_routine.start_routine)(thread_routine.args);
+        free(stack_addr);
+        result
+    }
+}
+
 #[no_mangle]
 pub extern "C" fn __wrap_pthread_create(
-    _: *mut c_void,
-    _: *const c_void,
-    _: extern "C" fn(*mut c_void) -> *mut c_void,
-    _: *mut c_void,
+    thread: *mut pthread_t,
+    attr: *const pthread_attr_t,
+    start_routine: extern "C" fn(*mut c_void) -> *mut c_void,
+    args: *mut c_void,
 ) -> c_int {
-    todo!("Allocate and set thread-local ShadowStack instance. No support for threads yet!");
+    unsafe {
+        let routine_args = &ThreadRoutine {
```

Comment:
The `let` isn't unsafe so can go outside of the `safe` block.

However, I'm not totally certain if passing a pointer to C code to something on the "Rust stack" is defined or undefined behaviour. You're going to have to do some reading up on UB with the C FFI in Rust to work that out I'm afraid!

Reply:
I think in both cases It is an undefined behaviour since `routine_args` might be freed by the time the created thread will run.

I switched to use [Box::into_raw](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.into_raw) and [Box::from_raw](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.from_raw) here 👉 ced50ed5a4004fe77b616dedcf25c07bea088d4d

Reply:
I *think* I see how this works, but it's not totally clear. We need to document clearly when we're passing around raw pointers like this so that we know whose role it is to free them: that means the return value(s) and parameter(s) will need `///` documenting.

Reply:
Updated 👉  bce94d1be08829164adf5879374b55f5062ffa01